tracking/reporter.cpp

```diff
@@ -0,0 +1,147 @@
+#include "reporter.hpp"
+
+#include "base/logging.hpp"
+#include "base/thread.hpp"
+#include "base/timer.hpp"
+
+#include "boost/circular_buffer.hpp"
+
+#include "platform/location.hpp"
+#include "platform/socket.hpp"
+
+#include "std/mutex.hpp"
+#include "std/vector.hpp"
+
+#include "tracking/connection.hpp"
+
+using namespace tracking;
+
+namespace
+{
+double constexpr kMinDelaySeconds = 1.0;
+double constexpr kReconnectDelaySeconds = 60.0;
+size_t constexpr kRealTimeBufferSize = 60;
+
+class WorkerImpl final : public Reporter::Worker
+{
+public:
+  WorkerImpl(unique_ptr<platform::Socket> socket, size_t pushDelayMs);
+  void Run();
+
+  // Worker overrides
+  void AddLocation(location::GpsInfo const & info);
+  void Stop();
+
+private:
+  void FetchInput();
+  bool SendPoints();
+
+  volatile bool m_stop = false;
+  Connection m_realtimeSender;
+  size_t m_pushDelayMs;
+  bool m_wasConnected = false;
+  double m_lastConnectTryTime = 0.0;
+  vector<tracking::DataPoint> m_input;
+  mutex m_inputMutex;
+  boost::circular_buffer<DataPoint> m_points;
+  double m_lastGpsTime = 0.0;
+};
+}  // namespace
+
+namespace tracking
+{
+// static
+const char Reporter::kAllowKey[] = "AllowStat";
+
+Reporter::Reporter(unique_ptr<platform::Socket> socket, size_t pushDelayMs)
+{
+  WorkerImpl * worker = new WorkerImpl(move(socket), pushDelayMs);
+  m_worker = worker;
+  threads::SimpleThread thread([worker]
+  {
+    worker->Run();
+    delete worker;
+  });
+  thread.detach();
```

Comment:
From my point of view, reporter should be instantiated once, as a field of Framework. And it will be destructed once, on the destruction of Framework, and it's not critical that destruction of Framework will be slightly blocked - you should cancel your worker on the reporter destruction and the worker should stop immediately. So, it's not a big delay, in any case.